### PR TITLE
raftstore-v2: fix potential data loss during server close

### LINKED ISSUE #14390: [Dynamic Regions] multiple rocksdb tikv panic for txn record found but not expected
### What version of TiKV are you using?

[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:86](http://lib.rs:86/)] ["Welcome to TiKV"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Release Version:   6.7.0-alpha"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Edition:           Community"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Git Commit Hash:   fc9cf096b3eddc48e2f1df09dd6d716198067410"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Git Commit Branch: heads/refs/tags/v6.7.0-alpha"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["UTC Build Time:    Unknown (env var does not exist when building)"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure"]
[2023/03/12 06:10:13.638 +00:00] [INFO] [[lib.rs:91](http://lib.rs:91/)] ["Profile:           dist_release"]
### Steps to reproduce

```
[2023/03/12 06:15:16.834 +00:00] [FATAL] [[lib.rs:497](http://lib.rs:497/)] ["txn record found but not expected: txn @440035367158808578"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tikv_util/src/lib.rs:496:18\n](http://github.com/pingcap/tikv/components/tikv_util/src/lib.rs:496:18/n)   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/[boxed.rs:2032](http://boxed.rs:2032/):9\n      std::panicking::rust_panic_with_hook\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:692](http://panicking.rs:692/):13\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:579](http://panicking.rs:579/):13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/[backtrace.rs:137](http://backtrace.rs:137/):18\n   4: rust_begin_unwind\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:575](http://panicking.rs:575/):5\n   5: core::panicking::panic_fmt\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/[panicking.rs:65](http://panicking.rs:65/):14\n   6: tikv::storage::txn::actions::check_txn_status::rollback_lock\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:158:13\n](http://github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:158:13/n)   7: tikv::storage::txn::actions::check_txn_status::check_txn_status_lock_exists\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:53:17\n](http://github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:53:17/n)      <tikv::storage::txn::commands::check_txn_status::CheckTxnStatus as tikv::storage::txn::commands::WriteCommand<S,L>>::process_write\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/commands/check_txn_status.rs:101:54\n](http://github.com/pingcap/tikv/src/storage/txn/commands/check_txn_status.rs:101:54/n)      tikv::storage::txn::commands::Command::process_write\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/commands/mod.rs:693:43\n](http://github.com/pingcap/tikv/src/storage/txn/commands/mod.rs:693:43/n)   8: tikv::storage::txn::scheduler::TxnScheduler<E,L>::process_write::{{closure}}::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1184:21\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1184:21/n)      tikv::storage::metrics::with_perf_context::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/metrics.rs:357:19\n](http://github.com/pingcap/tikv/src/storage/metrics.rs:357:19/n)      std::thread::local::LocalKey<T>::try_with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:446](http://local.rs:446/):16\n      std::thread::local::LocalKey<T>::with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:422](http://local.rs:422/):9\n      tikv::storage::metrics::with_perf_context\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/metrics.rs:348:5\n](http://github.com/pingcap/tikv/src/storage/metrics.rs:348:5/n)      tikv::storage::txn::scheduler::TxnScheduler<E,L>::process_write::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1183:17\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1183:17/n)   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      tikv::storage::txn::scheduler::TxnScheduler<E,L>::process::{{closure}}::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1097:68\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1097:68/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      <resource_metering::InTags<T> as core::future::future::Future>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/resource_metering/src/lib.rs:266:9\n](http://github.com/pingcap/tikv/components/resource_metering/src/lib.rs:266:9/n)      tikv::storage::txn::scheduler::TxnScheduler<E,L>::process::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1110:9\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1110:9/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      tikv::storage::txn::scheduler::TxnScheduler<E,L>::execute::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:715:54\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:715:54/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n  10: tikv::storage::txn::sched_pool::SchedPool::spawn::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/sched_pool.rs:153:30\n](http://github.com/pingcap/tikv/src/storage/txn/sched_pool.rs:153:30/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      <resource_control::future::ControlledFuture<F> as core::future::future::Future>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/resource_control/src/future.rs:39:19\n](http://github.com/pingcap/tikv/components/resource_control/src/future.rs:39:19/n)      <tracker::tls::TrackedFuture<F> as core::future::future::Future>::poll::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tracker/src/tls.rs:64:23\n](http://github.com/pingcap/tikv/components/tracker/src/tls.rs:64:23/n)      std::thread::local::LocalKey<T>::try_with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:446](http://local.rs:446/):16\n      std::thread::local::LocalKey<T>::with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:422](http://local.rs:422/):9\n      <tracker::tls::TrackedFuture<F> as core::future::future::Future>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tracker/src/tls.rs:62:9\n](http://github.com/pingcap/tikv/components/tracker/src/tls.rs:62:9/n)      tikv_util::yatp_pool::future_pool::PoolInner::spawn::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/future_pool.rs:167:27\n](http://github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/future_pool.rs:167:27/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      yatp::task::future::RawTask<F>::poll\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/task/[future.rs:59](http://future.rs:59/):9\n  11: yatp::task::future::TaskCell::poll\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/task/[future.rs:103](http://future.rs:103/):9\n      <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/task/[future.rs:387](http://future.rs:387/):20\n  12: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:193:24\n](http://github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:193:24/n)  13: <yatp::queue::priority::PriorityRunner<R> as yatp::pool::runner::Runner>::handle\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/queue/[priority.rs:191](http://priority.rs:191/):19\n      yatp::pool::worker::WorkerThread<T,R>::run\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/pool/[worker.rs:48](http://worker.rs:48/):13\n      yatp::pool::builder::LazyBuilder<T>::build::{{closure}}\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/pool/[builder.rs:114](http://builder.rs:114/):25\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/[backtrace.rs:121](http://backtrace.rs:121/):18\n  14: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[mod.rs:551](http://mod.rs:551/):17\n      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/[unwind_safe.rs:271](http://unwind_safe.rs:271/):9\n      std::panicking::try::do_call\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:483](http://panicking.rs:483/):40\n      std::panicking::try\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:447](http://panicking.rs:447/):19\n      std::panic::catch_unwind\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panic.rs:137](http://panic.rs:137/):14\n      std::thread::Builder::spawn_unchecked_::{{closure}}\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[mod.rs:550](http://mod.rs:550/):30\n      core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/[function.rs:513](http://function.rs:513/):5\n  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/[boxed.rs:2000](http://boxed.rs:2000/):9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/[boxed.rs:2000](http://boxed.rs:2000/):9\n      std::sys::unix::thread::Thread::new::thread_start\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/[thread.rs:108](http://thread.rs:108/):17\n  16: start_thread\n  17: clone\n"] [location=/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:158](http://github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:158)] [thread_name=sched-worker-pool-5]
store/TiDB master sequential table-cache auto-retry auto-retry-limit :default select FOR UPDATE txn-mode pessimistic isolation :repeatable-read/20230312T060931.000Z/node-4.node-peer.jepsen-tps-1656503-1-46/kv.log:[2023/03/12 06:15:16.834 +00:00] [FATAL] [[lib.rs:497](http://lib.rs:497/)] ["txn record found but not expected: txn @440035367355678736"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tikv_util/src/lib.rs:496:18\n](http://github.com/pingcap/tikv/components/tikv_util/src/lib.rs:496:18/n)   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/[boxed.rs:2032](http://boxed.rs:2032/):9\n      std::panicking::rust_panic_with_hook\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:692](http://panicking.rs:692/):13\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:579](http://panicking.rs:579/):13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/[backtrace.rs:137](http://backtrace.rs:137/):18\n   4: rust_begin_unwind\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:575](http://panicking.rs:575/):5\n   5: core::panicking::panic_fmt\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/[panicking.rs:65](http://panicking.rs:65/):14\n   6: tikv::storage::txn::actions::check_txn_status::rollback_lock\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:158:13\n](http://github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:158:13/n)   7: tikv::storage::txn::actions::check_txn_status::check_txn_status_lock_exists\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:53:17\n](http://github.com/pingcap/tikv/src/storage/txn/actions/check_txn_status.rs:53:17/n)      <tikv::storage::txn::commands::check_txn_status::CheckTxnStatus as tikv::storage::txn::commands::WriteCommand<S,L>>::process_write\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/commands/check_txn_status.rs:101:54\n](http://github.com/pingcap/tikv/src/storage/txn/commands/check_txn_status.rs:101:54/n)      tikv::storage::txn::commands::Command::process_write\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/commands/mod.rs:693:43\n](http://github.com/pingcap/tikv/src/storage/txn/commands/mod.rs:693:43/n)   8: tikv::storage::txn::scheduler::TxnScheduler<E,L>::process_write::{{closure}}::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1184:21\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1184:21/n)      tikv::storage::metrics::with_perf_context::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/metrics.rs:357:19\n](http://github.com/pingcap/tikv/src/storage/metrics.rs:357:19/n)      std::thread::local::LocalKey<T>::try_with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:446](http://local.rs:446/):16\n      std::thread::local::LocalKey<T>::with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:422](http://local.rs:422/):9\n      tikv::storage::metrics::with_perf_context\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/metrics.rs:348:5\n](http://github.com/pingcap/tikv/src/storage/metrics.rs:348:5/n)      tikv::storage::txn::scheduler::TxnScheduler<E,L>::process_write::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1183:17\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1183:17/n)   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      tikv::storage::txn::scheduler::TxnScheduler<E,L>::process::{{closure}}::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1097:68\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1097:68/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      <resource_metering::InTags<T> as core::future::future::Future>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/resource_metering/src/lib.rs:266:9\n](http://github.com/pingcap/tikv/components/resource_metering/src/lib.rs:266:9/n)      tikv::storage::txn::scheduler::TxnScheduler<E,L>::process::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1110:9\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:1110:9/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      tikv::storage::txn::scheduler::TxnScheduler<E,L>::execute::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/scheduler.rs:715:54\n](http://github.com/pingcap/tikv/src/storage/txn/scheduler.rs:715:54/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n  10: tikv::storage::txn::sched_pool::SchedPool::spawn::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/src/storage/txn/sched_pool.rs:153:30\n](http://github.com/pingcap/tikv/src/storage/txn/sched_pool.rs:153:30/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      <resource_control::future::ControlledFuture<F> as core::future::future::Future>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/resource_control/src/future.rs:39:19\n](http://github.com/pingcap/tikv/components/resource_control/src/future.rs:39:19/n)      <tracker::tls::TrackedFuture<F> as core::future::future::Future>::poll::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tracker/src/tls.rs:64:23\n](http://github.com/pingcap/tikv/components/tracker/src/tls.rs:64:23/n)      std::thread::local::LocalKey<T>::try_with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:446](http://local.rs:446/):16\n      std::thread::local::LocalKey<T>::with\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[local.rs:422](http://local.rs:422/):9\n      <tracker::tls::TrackedFuture<F> as core::future::future::Future>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tracker/src/tls.rs:62:9\n](http://github.com/pingcap/tikv/components/tracker/src/tls.rs:62:9/n)      tikv_util::yatp_pool::future_pool::PoolInner::spawn::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/future_pool.rs:167:27\n](http://github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/future_pool.rs:167:27/n)      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/[mod.rs:91](http://mod.rs:91/):19\n      yatp::task::future::RawTask<F>::poll\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/task/[future.rs:59](http://future.rs:59/):9\n  11: yatp::task::future::TaskCell::poll\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/task/[future.rs:103](http://future.rs:103/):9\n      <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/task/[future.rs:387](http://future.rs:387/):20\n  12: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle\n             at /home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:193:24\n](http://github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:193:24/n)  13: <yatp::queue::priority::PriorityRunner<R> as yatp::pool::runner::Runner>::handle\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/queue/[priority.rs:191](http://priority.rs:191/):19\n      yatp::pool::worker::WorkerThread<T,R>::run\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/pool/[worker.rs:48](http://worker.rs:48/):13\n      yatp::pool::builder::LazyBuilder<T>::build::{{closure}}\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/7ed2529/src/pool/[builder.rs:114](http://builder.rs:114/):25\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/[backtrace.rs:121](http://backtrace.rs:121/):18\n  14: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/[mod.rs:551](http://mod.rs:551/):17\n      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/[unwind_safe.rs:271](http://unwind_safe.rs:271/):9\n      std::panicking::try::do_call\n             at /rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/[panicking.rs:483](http://panicking.rs:483/):40\n      std::panicking::try\n             at /rust/t
```

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #14390

What's Changed:

```commit-message
fix potential data loss during server close
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: tests/integrations/raftstore/test_bootstrap.rs [hunk 1]
```diff
@@ -249,3 +250,68 @@ fn test_flush_before_stop() {
...
+// 2. on_applied_res => lockcf's last_modified = 6
+// 3. flush lock cf => lockcf's flushed_index = 6
+// 4. batch {unlock `k`, write `k`} with index 7
+//
+// flush-before-close:
+// 5. pick write cf to flush => writecf's flushed_index = 7
+//
+// 6. maybe_advance_admin_flushed(7): as lockcf's last_modified = flushed_index,
+// it will not block advancing admin index
+// 7. admin index 7 is persisted. => we may loss `unlock k`
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_bootstrap.rs:267]: on_apply_res  is not called yet.

[the remark this answers is not in the document]
Comment (tabokie) [reply tests/integrations/raftstore/test_bootstrap.rs:267]: Okay, I think you can add "last_modified is updated in store but RocksDB is modified in apply."

[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_bootstrap.rs:267]: Fine.

File: components/raftstore-v2/src/operation/ready/apply_trace.rs [hunk 2]
```diff
@@ -627,21 +631,21 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
                 let flushed_indexes = flush_state.as_ref().flushed_index();
                 for i in 0..flushed_indexes.len() {
                     let flush_index = flushed_indexes[i].load(Ordering::SeqCst);
                     let cf = offset_to_cf(i);
                     apply_trace.on_flush(cf, flush_index);
-                    max_flush_index = u64::max(max_flush_index, flush_index);
                 }
 
-                apply_trace.maybe_advance_admin_flushed(max_flush_index);
+                apply_trace.maybe_advance_admin_flushed(applied_index);
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/ready/apply_trace.rs:642]: Question:  we just flushed one cf (the oldest cf), why can the flushed index be set to applied_index, as other CFs have not been flushed yet.

File: components/raftstore-v2/src/operation/ready/apply_trace.rs [hunk 3]
```diff
@@ -627,21 +631,21 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
                 let admin_flush = apply_trace.admin.flushed;
                 apply_trace.persisted_applied = admin_flush;
 
-                if self.storage().estimate_replay_count() < flush_threshold {
+                if self.storage().estimate_replay_count() < flush_threshold
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/ready/apply_trace.rs:646]: Why "if self.storage().estimate_replay_count() < flush_threshold". is it checked in L622?

Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/ready/apply_trace.rs:646]: why not moving this flush just below the loop, as long as there's a flush.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/ready/apply_trace.rs]: Good suggestion and done.

**Review by tonyxuqqi — APPROVED** [review]

**Review by tabokie — APPROVED** [review]

**Pull request merged**